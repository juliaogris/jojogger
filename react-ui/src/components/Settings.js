import React, { Component, PropTypes } from 'react'
import { updateUser } from '../util/api'
import { getEmailError, getPasswordError } from '../util/funcs'
import TickButton from './elements/TickButton'

export default class Settings extends Component {
  constructor (props) {
    super(props)
    this.state = {
      error: null,
      email: this.props.user.email,
      password: ''
    }
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleInputChange = this.handleInputChange.bind(this)
  }

  handleSubmit (event) {
    event.preventDefault()
    const authedUser = this.props.user
    const { email, password } = this.state
    let error = getEmailError(email)
    const updatedUser = { email, id: authedUser.id }
    if (password) {
      updatedUser.password = password
      error = error || getPasswordError(password)
    }

    if (error) {
      this.setState({ error })
      return
    }
    updateUser(authedUser, updatedUser)
      .then(() => {
        this.props.signOut()
      })
      .catch((error) => {
        this.setState({ error })
      })
  }

  handleInputChange (event) {
    const { name, value } = event.target
    this.setState({ [name]: value, error: null })
  }

  render () {
    const { email, password, error } = this.state

    return (
      <form className='jojog-form' action={this.handleSubmit}>
        {error && <div className='settings-error'>{error.message}</div>}
        <label>email</label>
        <input
          placeholder='email@example.com'
          type='email'
          name='email'
          value={email}
          onChange={this.handleInputChange}
        />
        <label>password</label>
        <input
          placeholder='your password'
          type='password'
          name='password'
          value={password}
          onChange={this.handleInputChange}
          />
        <TickButton onClick={this.handleSubmit} />
      </form>
    )
  }
}
Settings.propTypes = {
  user: PropTypes.object.isRequired,
  signOut: PropTypes.func.isRequired
}
