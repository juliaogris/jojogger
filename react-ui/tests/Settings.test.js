/* global it expect jest */
import React from 'react'
import ReactDOM from 'react-dom'
import Settings from '../src/components/Settings'
import renderer from 'react-test-renderer'
import ReactTestUtils from 'react-addons-test-utils'

jest.unmock('../src/util/api')
const api = require('../src/util/api')

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Settings
    user={{ email: 'example@x.co' }}
    signOut={() => {}}
    />, div)
})

it('renders "Settings" and matches snapshot', () => {
  const tree = renderer.create(<Settings
    user={{ email: 'example@x.co' }}
    signOut={() => {}}
    />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('calls RestAPI.updateUser and signOut on submit click', () => {
  const signOut = jest.fn()
  const settings = ReactTestUtils.renderIntoDocument(<Settings
    user={{ email: 'example@x.co', id: 'myIdXyz' }}
    signOut={signOut}
  />)
  settings.setState({ email: 'j@x.co', password: 'pa$$word' })
  api.updateUser = jest.fn(() => ({
    then: (func) => {
      func()
      return { catch: () => {} }
    }
  }))
  settings.handleSubmit()
  expect(signOut).toHaveBeenCalledTimes(1)
  expect(api.updateUser).toHaveBeenCalledTimes(1)
  expect(api.updateUser).toHaveBeenCalledWith(
    { email: 'example@x.co', id: 'myIdXyz' },
    { email: 'j@x.co', password: 'pa$$word', id: 'myIdXyz' })
})

it('delete Account with prior confirmation prompt', () => {
  const signOut = jest.fn()
  const user = { email: 'example@x.co', id: 'myIdXyz' }
  const settings = ReactTestUtils.renderIntoDocument(<Settings
    user={user}
    signOut={signOut}
  />)
  let buttons = ReactTestUtils.scryRenderedDOMComponentsWithTag(settings, 'span')
  expect(buttons.length).toEqual(1)
  const deleteButton = buttons[0]
  ReactTestUtils.Simulate.click(deleteButton)
  expect(settings.state.deleteConfirm).toBeTruthy()
  buttons = ReactTestUtils.scryRenderedDOMComponentsWithTag(settings, 'button')
  expect(buttons.length).toEqual(2)

  const yesButton = buttons[1]
  api.deleteUser = jest.fn(() => ({
    then: (func) => {
      func()
      return { catch: () => {} }
    }
  }))
  ReactTestUtils.Simulate.click(yesButton)
  expect(signOut).toHaveBeenCalledTimes(1)
  expect(api.deleteUser).toHaveBeenCalledTimes(1)
  expect(api.deleteUser).toHaveBeenCalledWith(user, user)
})
