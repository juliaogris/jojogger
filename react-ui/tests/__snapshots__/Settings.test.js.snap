exports[`test renders "Settings" and matches snapshot 1`] = `
<form
  action={[Function]}
  className="jojog-form">
  <h2
    className="settings-heading">
    Settings
  </h2>
  <label>
    Update email
  </label>
  <input
    name="email"
    onChange={[Function]}
    placeholder="email@example.com"
    type="email"
    value="example@x.co" />
  <label>
    Update password
  </label>
  <input
    name="password"
    onChange={[Function]}
    placeholder="your password"
    type="password"
    value="" />
  <div
    className="delete-row">
    <span
      className="delete-button"
      onClick={[Function]}>
      Delete this 
      Account
    </span>
  </div>
  <div
    className="tick-button-wrap">
    <button
      className="tick-button"
      onClick={[Function]}
      type="submit">
      <svg
        className="add-icon"
        height="100%"
        viewBox="0 -1  41 42"
        width="100%"
        xmlns="http://www.w3.org/2000/svg">
        <g
          fill="none"
          fillRule="evenodd"
          stroke="none"
          strokeWidth="1">
          <polygon
            fill="white"
            fillRule="nonzero"
            points="14.5299498 13.3700502 15.9299498 11.9700502 26.9458691 22.4458691 20.9458691 28.4458691 19.5458691 27.0458691 24.1458691 22.4458691"
            transform="translate(20.737909, 20.207960) rotate(90.000000) translate(-20.737909, -20.207960) " />
        </g>
      </svg>
    </button>
  </div>
</form>
`;
